p2p/tpu.md

```diff
@@ -0,0 +1,247 @@
+
+TPU/QUIC implements fragmentation by splitting a QUIC stream across multiple packets.
+The payload of each QUIC packet should maximize space up to the MTU.
+
+Peers must avoid sending QUIC traffic using IPv4/IPv6 fragmentation.
+Incoming QUIC traffic indicating IP fragmentation may be ignored.
+
+**Example QUIC frames**
+
+_First UDP packet_
+
+```
+STREAM Frame {
+  Prefix (5) = 0b00001
+  Offset Present (1) = 0   # Offset is absent for first fragment
+  Length Present (1) = 0   # Length is absent, as fragment frames always fill packet
+  Fin (1) = 0
+
+  Stream ID (i)
+  Stream Data (..)
+}
+```
+
+_Second UDP packet_
+
+```
+STREAM Frame {
+  Prefix (5) = 0b00001
+  Offset Present (1) = 1
+  Length Present (1) = 0
+  Fin (1) = 0
+
+  Stream ID (i)
+  [Offset (i)]
+  Stream Data (..)
+}
+```
+
+_Last UDP packet_
+
+```
+STREAM Frame {
+  Prefix (5) = 0b00001
+  Offset Present (1) = 1
+  Length Present (1) = 0
+  Fin (1) = 1             # Terminate stream
+
+  Stream ID (i)
+  [Offset (i)]
+  Stream Data (..)
+}
+```
+
+#### Acknowledgement
+	
+`STREAM` frames sent by clients are ack-eliciting.
+Refer to [RFC 9000 Section 13.2.1] for obligations on sending `ACK` frames.
+
+It is permitted to coalesce multiple `ACK` frames to reduce packet count.
+
+#### Stream Limits
+
+The server should send [`MAX_STREAM_DATA`] frames for recently created fragmented tx streams.
+The client should expect arbitrarily receive quotas for any stream.
+
+Note that stream IDs are incremental (ignoring the two least significant bits).
+Upon opening a stream with an out-of-order ID, all preceding streams are also created.
+
+In this case, the server may proactively send multiple [`MAX_STREAM_DATA`] frames for implicitly created streams.
+
+  [RFC 9000 Section 13.2.1]: https://www.rfc-editor.org/rfc/rfc9000.html#name-sending-ack-frames
+  [`MAX_STREAM_DATA`]: https://www.rfc-editor.org/rfc/rfc9000.html#name-max_stream_data-frames
+
+#### Aborting
+
+The server must terminate the QUIC connection in case of the following events:
+
+  - Submission of invalid transactions, specifically:
+    - Txs that fail deserialization
+    - Txs that fail parameter sanitization
+    - Txs that fail signature verification
+  - Transmission of QUIC datagrams, defined by [RFC 9221]
+  - Attempted creation of QUIC bidirectional streams (already forbidden by quota)
+  - All other cases where the QUIC RFC dictates connection termination.
+
+  [RFC 9221]: https://www.rfc-editor.org/rfc/rfc9221.html
+
+## TPU/UDP protocol
+
+TPU/UDP is a simple datagram-oriented network protocol.
+
+Each datagram carries a single serialized transaction without headers or padding.
+
+The protocol is strictly unidirectional. The receiver should therefore ignore the source IP address and UDP port.
+
+It is being deprecated in favor of TPU/QUIC, which adds authentication, confidentiality, and congestion control.
+
+TPU/UDP traffic cannot be multiplexed with other protocols over the same destination port.
```

Comment:
@ryleung-solana This is a fair point and we should include this info somewhere in this repo. However, I would be hesitant to put this into the TPU/QUIC spec, since this concerns implementation/usage of the protocol, not the protocol itself.

Therefore I would put this in a separate page that describes the context of the TPU/QUIC deployment in the validator network specifically.

IMO, if someone wants to deploy TPU/QUIC outside of the validator network (e.g. private network infra), then they should not have to abide by this restriction. If the spec stays neutral, then they can still be compliant with TPU/QUIC without using gossip which allows reusing existing client libraries.

Reply:
Fair point.